### PR TITLE
Cameras and lights now take power from adjacent walls if placed in external areas without power

### LINKED ISSUE #68952: Camera saves you from Ice Storm

### PR DESCRIPTION
**Pull request by MMMiracles**
## About The Pull Request
Adds a proc to machinery that can be used to find a valid powered area for a reference if just checking the turf its on isn't enough.
- checks turf its sitting on if always_powered is false
- if true, checks adjacent wall based on machine's direction (using the right direction for light/cameras is important now actually)
- if no adjacent wall/adjacent wall doesn't have a good area, defaults back to original turf's area

Adds this proc in place of locations where cameras and lighting fixtures checked for their areas for power/lightswitch requirements.

Closes #68952
## Why It's Good For The Game
Removes the need to do the strange work-around where you place a singular powered area and causing issues if the map has outside weather.

## Changelog

:cl: MMMiracles
fix: Cameras and Lights on external hulls are no long safe havens from the rough reality of the outdoors in harsh weather situations.
/:cl:

### CODE REVIEW

File: code/game/machinery/_machinery.dm [hunk 1]
```diff
@@ -485,6 +485,21 @@
 
 	return TRUE
 
+///Get a valid powered area to reference for power use, mainly for wall-mounted machinery that isn't always mapped directly in a powered location.
+/obj/machinery/proc/get_room_area(area/machine_room)
+	var/area/machine_area = get_area(src)
+	if(machine_area.always_unpowered == FALSE)///check our loc first to see if its a powered area
+		machine_room= machine_area
```
Comment (san7890) [inline code/game/machinery/_machinery.dm:492]:
```suggestion
		machine_room = machine_area
```

looks better

File: code/modules/power/lighting/light.dm [hunk 2]
```diff
@@ -133,7 +134,7 @@
 	if(!on || status != LIGHT_OK)
 		return
 
-	var/area/local_area = get_area(src)
+	var/area/local_area =get_room_area(src)
```
Comment (san7890) [inline code/modules/power/lighting/light.dm:137]:
```suggestion
	var/area/local_area = get_room_area(src)
```
style

File: code/game/machinery/_machinery.dm [hunk 3]
```diff
@@ -485,6 +485,21 @@
 
 	return TRUE
 
+///Get a valid powered area to reference for power use, mainly for wall-mounted machinery that isn't always mapped directly in a powered location.
+/obj/machinery/proc/get_room_area(area/machine_room)
+	var/area/machine_area = get_area(src)
+	if(machine_area.always_unpowered == FALSE)///check our loc first to see if its a powered area
```
Comment (JohnFulpWillard) [inline code/game/machinery/_machinery.dm:491]:
```suggestion
	if(!machine_area.always_unpowered) //check our loc first to see if its a powered area
```

File: code/game/machinery/_machinery.dm [hunk 4]
```diff
@@ -485,6 +485,21 @@
...
+		machine_room = machine_area
+		return machine_room
+	var/turf/mounted_wall = get_step(src,dir)
+	if(mounted_wall && istype(mounted_wall, /turf/closed))
+		var/area/wall_area = get_area(mounted_wall)
+		if(wall_area.always_unpowered == FALSE)///loc area wasn't good, checking adjacent wall for a good area to use
```
Comment (JohnFulpWillard) [inline code/game/machinery/_machinery.dm:497]:
```suggestion
		if(!wall_area.always_unpowered) //loc area wasn't good, checking adjacent wall for a good area to use
```

File: code/game/machinery/_machinery.dm [hunk 5]
```diff
@@ -485,6 +485,21 @@
...
+			machine_room = wall_area
+			return machine_room
+	machine_room = machine_area///couldn't find a proper powered area on loc or adjacent wall, defaulting back to loc and blaming mappers
```
Comment (JohnFulpWillard) [inline code/game/machinery/_machinery.dm:500]:
```suggestion
	machine_room = machine_area //couldn't find a proper powered area on loc or adjacent wall, defaulting back to loc and blaming mappers
```

File: code/game/machinery/_machinery.dm [hunk 6]
```diff
@@ -485,6 +485,21 @@
...
+	return machine_room
```
Comment (JohnFulpWillard) [inline code/game/machinery/_machinery.dm:501]: Just an idea if you want to reduce repeating yourself a little, replacing the returns with ``else`` so you can just have one ``return machine_room`` at the very end.

```suggestion
	var/area/machine_area = get_area(src)
	if(machine_area.always_unpowered == FALSE)///check our loc first to see if its a powered area
		machine_room = machine_area
	var/turf/mounted_wall = get_step(src,dir)
	else if(mounted_wall && istype(mounted_wall, /turf/closed))
		var/area/wall_area = get_area(mounted_wall)
		if(wall_area.always_unpowered == FALSE)///loc area wasn't good, checking adjacent wall for a good area to use
			machine_room = wall_area
	else
		machine_room = machine_area///couldn't find a proper powered area on loc or adjacent wall, defaulting back to loc and blaming mappers
	return machine_room
```

File: code/game/machinery/_machinery.dm [hunk 7]
```diff
@@ -485,6 +485,21 @@
...
+		machine_room = machine_area
+		return machine_room
+	var/turf/mounted_wall = get_step(src,dir)
+	if (mounted_wall && istype(mounted_wall, /turf/closed))
+		var/area/wall_area = get_area(mounted_wall)
+		if(!wall_area.always_unpowered )///loc area wasn't good, checking adjacent wall for a good area to use
+			machine_room = wall_area
+			return machine_room
+	machine_room = machine_area ///couldn't find a proper powered area on loc or adjacent wall, defaulting back to loc and blaming mappers
+	return machine_room
```
Comment (JohnFulpWillard) [inline code/game/machinery/_machinery.dm:501]:
```suggestion
	var/area/machine_area = get_area(src)
	if(!machine_area.always_unpowered) ///check our loc first to see if its a powered area
		machine_room = machine_area
		return machine_room
	var/turf/mounted_wall = get_step(src,dir)
	if (mounted_wall && istype(mounted_wall, /turf/closed))
		var/area/wall_area = get_area(mounted_wall)
		if(!wall_area.always_unpowered) //loc area wasn't good, checking adjacent wall for a good area to use
			machine_room = wall_area
			return machine_room
	machine_room = machine_area ///couldn't find a proper powered area on loc or adjacent wall, defaulting back to loc and blaming mappers
	return machine_room
```

File: code/game/machinery/_machinery.dm [hunk 8]
```diff
@@ -485,6 +485,21 @@
...
+///Get a valid powered area to reference for power use, mainly for wall-mounted machinery that isn't always mapped directly in a powered location.
+/obj/machinery/proc/get_room_area(area/machine_room)
+	var/area/machine_area = get_area(src)
+	if(!machine_area.always_unpowered) ///check our loc first to see if its a powered area
+		machine_room = machine_area
+		return machine_room
+	var/turf/mounted_wall = get_step(src,dir)
+	if (mounted_wall && istype(mounted_wall, /turf/closed))
+		var/area/wall_area = get_area(mounted_wall)
+		if(!wall_area.always_unpowered )///loc area wasn't good, checking adjacent wall for a good area to use
```
Comment (JohnFulpWillard) [inline code/game/machinery/_machinery.dm:497]:
```suggestion
		if(!wall_area.always_unpowered) //loc area wasn't good, checking adjacent wall for a good area to use
```

**Pull request merged**